### PR TITLE
Add char matrices for Valencia

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

~Dependent on previous PRs. Please only review "Add characteristic matrices for Valencia" commit.~

~Depends on #1958.~ 

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 1]
```diff
@@ -87,20 +91,297 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+                          const tnsr::II<double, Dim>& inv_spatial_metric,
+                          const Scalar<double>& det_spatial_metric,
+                          const tnsr::i<double, Dim>& unit_normal) noexcept {
+  Matrix result(Dim + 2, Dim + 2);
+
+  const auto spatial_velocity_oneform =
+      raise_or_lower_index(spatial_velocity, spatial_metric);
+
+  // save allocation by storing normal velocity here
+  double prefactor = get(dot_product(unit_normal, spatial_velocity));
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: Doubles aren't allocated — they're placed on the stack and super cheap.

So I would strongly suggest favoring readability and stricter use of `const` by writing `const double normal_velocity = ...` here, `const double velocity_squared = ...` below, and so on elsewhere.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 2]
```diff
@@ -87,20 +91,297 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+      get(lorentz_factor) *
+      (get(specific_internal_energy) + get(pressure) / get(rest_mass_density) +
+       get(lorentz_factor) * h_w_minus_one / (get(lorentz_factor) + 1.0));
+
+  const double specific_enthalpy_times_lorentz_factor =
+      get(specific_enthalpy) * get(lorentz_factor);
+
+  prefactor *= specific_enthalpy_times_lorentz_factor;
+
+  // eigenvalue = lapse * Lambda_minus - normal_shift
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: To make it more clear how this comment relates to the code, I suggest to rephrase this comment as "Set the eigenvector that has eigenvalue = ...", here and in similar comments below.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 3]
```diff
@@ -74,8 +75,131 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
     const tnsr::i<DataVector, Dim>& normal) noexcept;
 // @}
 
-namespace Tags {
+/*!
+ * \brief Right eigenvectors of the Valencia formulation
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:79]: Possibly in the brief, but ideally also in the main body, say what this means: these vectors let you transform between the conserved and characteristic variables of the system.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 4]
```diff
@@ -74,8 +75,131 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+ *
+ * The right eigenvectors correspond to the columns of the matrix
+ * \f$\mathbf R = [R_-\, R_1\, R_2\, R_3\, R_+]\f$, where
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp]: For most users of the code, how this matrix is to be used will be more important than the precise expressions. I suggest that before giving the full mathematical expressions, you document the usage of "char = left * cons" and "cons = right * char", being clear about the ordering of the cons and char variables.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 5]
```diff
@@ -74,8 +75,131 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+ * Wv_{(1,2)} \\
+ * W\left(2hW - 1\right)v_{(1,2)} \\
+ * h\left(t_{(1,2)i} + 2W^2v_{(1,2)}v_i \right)
+ * \end{array}\right],\qquad
+ * R_3 = \left[\begin{array}{c}
+ * K \\
+ * K(hW - 1) - hWc_s^2 \\
+ * hW(K - c_s^2)v_i
+ * \end{array}\right].
+ * \f}
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:106]: Add reference

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 6]
```diff
@@ -74,8 +75,131 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+                          const Scalar<double>& kappa_over_density,
+                          const Scalar<double>& sound_speed_squared,
+                          const Scalar<double>& lorentz_factor,
+                          const tnsr::ii<double, Dim>& spatial_metric,
+                          const tnsr::II<double, Dim>& inv_spatial_metric,
+                          const Scalar<double>& det_spatial_metric,
+                          const tnsr::i<double, Dim>& unit_normal) noexcept;
 
+/*!
+ * \brief Left eigenvectors of the Valencia formulation
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:150]: All the same comments as for the right case.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp [hunk 7]
```diff
@@ -127,16 +133,105 @@ void test_with_normal_along_coordinate_axes(
...
+      make_not_null(&generator), used_for_size);
+  const auto det_and_inv_metric = determinant_and_inverse(spatial_metric);
+  const auto lorentz_factor = hydro::TestHelpers::random_lorentz_factor(
+      make_not_null(&generator), used_for_size);
+  const auto spatial_velocity = hydro::TestHelpers::random_velocity(
+      make_not_null(&generator), lorentz_factor, spatial_metric);
+
+  // Make consistent set of thermodynamic variables, along with their
+  // derived quantities for a given equation of state.
+  const EquationsOfState::IdealFluid<true> equation_of_state{5.0 / 3.0};
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I recommend testing with a Polytrope EoS too.

For this equation of state `kappa` goes to zero, so it's good to have a test that checks the choice of left/right vectors has no divisions by kappa. I did this wrong in the Newtonian case, and am currently working to fix it...

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp [hunk 8]
```diff
@@ -127,16 +133,105 @@ void test_with_normal_along_coordinate_axes(
...
+        det_and_inv_metric.first, basis_form);
+
+    left_matrix = RelativisticEuler::Valencia::left_eigenvectors(
+        rest_mass_density, spatial_velocity, specific_internal_energy, pressure,
+        specific_enthalpy, kappa_over_density, sound_speed_squared,
+        lorentz_factor, spatial_metric, det_and_inv_metric.second,
+        det_and_inv_metric.first, basis_form);
+
+    check_identity(left_matrix, right_matrix);
+  }
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Please add test that "right * speeds * left = d(flux)/du".

(Or is this far too difficult in the relativistic case? I am concerned that the current test does not check that these matrices are giving the correct transform... only that the two matrices are inverses of each other)

**Review by fmahebert — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp (hunk 1 above)
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: Back when I started working on this, I had these functions defined for a type different from `Matrix`, and the functions were templated in `DataType`, so allocations were important. Then I adopted the convention you implemented for the Newtonian case, which only works for `double`s, but decided to keep the code in case we decided to make it work for `DataVector`s as well. I can rework some terms a bit if you do not foresee any use for the matrices defined for `DataVector`s. What do you think?

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp (hunk 8 above)
Comment (carmaza) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I've worked out some terms and I was planning on adding that test too. The calculation is lengthy, so please do not wait if you need this code in develop.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp (hunk 1 above)
Comment (fmahebert) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: I think your question is equivalent to this one: will we need to compute characteristic fields in a _pointwise_ way, i.e., with a characteristic transformation that is different at every grid point? I don't believe I've seen algorithms that expect this (aside from, e.g., the upwind flux for GH which we wouldn't want to use for hydro anyway). So my suspicion is that it is unlikely we would need this these matrices in terms of DataVectors... Is this consistent with your thoughts?

Edit to clarify: if you're aware of any cases that might require the char transformation to be computed pointwise, then I have no problem with allowing it in the code :)

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp (hunk 8 above)
Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Okay, good to hear. I do not plan to use this transformation just yet (and can always get it from this branch), so I'm fine with waiting to merge this PR until you finish working it out. But if you prefer to do this as two steps with two PRs, that's okay with me too.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp (hunk 7 above)
Comment (teukolsky) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: Actually, in the relativistic case kappa = rho c_s^2 in this case (rho = rest mass). I think the difference is because of including rest mass in the energy density. The eigenvectors should be fine, the way I wrote them. It would be good to check numerically!

Comment (fmahebert) [reply tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: I haven't looked into this yet, but our `PolytropicFluid` equation of state sets `kappa_times_p_over_rho_squared` to 0 whether relativistic or not.

**Comment by carmaza** [comment]
With #1959 just merged, this PR should be easier to re-review. The fixups after @fmahebert 's review were a bit long so I split them into different commits in order to make them easier to read. Everything should be squashed after approving them.

This PR still depends on #1958 , which was already reviewed by @fmahebert but it still needs a second reviewer.

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp [hunk 9]
```diff
@@ -116,6 +117,91 @@ void test_characteristic_speeds(const DataVector& used_for_size) noexcept {
...
+  const double v_n = get(dot_product(spatial_velocity, normal));
+  const double v_squared = get(spatial_velocity_squared);
+  const double c_squared = get(sound_speed_squared);
+  const double prefactor = 1.0 / (1.0 - v_squared * c_squared);
+  const double first_term = prefactor * (1.0 - c_squared);
+  const double second_term =
+      prefactor *
+      sqrt(c_squared * (1.0 - v_squared) *
+           (1.0 - v_squared * c_squared - square(v_n) * (1.0 - c_squared)));
+  Matrix result(Dim + 2, Dim + 2, 0.0);
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp:140]: [nitpick] I suggest adding a blank line above this for readability

File: tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp [hunk 10]
```diff
@@ -116,6 +117,91 @@ void test_characteristic_speeds(const DataVector& used_for_size) noexcept {
...
+Matrix expected_flux_jacobian(
+    const tnsr::I<double, Dim>& spatial_velocity,
+    const tnsr::i<double, Dim>& spatial_velocity_oneform,
+    const Scalar<double>& lorentz_factor,
+    const Scalar<double>& specific_enthalpy,
+    const Scalar<double>& sound_speed_squared,
+    const Scalar<double>& kappa_over_density,
+    const tnsr::i<double, Dim>& normal,
+    const tnsr::I<double, Dim>& normal_vector) noexcept {
+  Matrix jacobian(Dim + 2, Dim + 2, 0.0);
```
Comment (fmahebert) [inline tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/Test_Characteristics.cpp]: [nitpick] I suggest moving this down to line ~167 where it is first used.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 11]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+                          const Scalar<double>& pressure,
+                          const Scalar<double>& specific_enthalpy,
+                          const Scalar<double>& kappa_over_density,
+                          const Scalar<double>& sound_speed_squared,
+                          const Scalar<double>& lorentz_factor,
+                          const tnsr::ii<double, Dim>& spatial_metric,
+                          const tnsr::II<double, Dim>& inv_spatial_metric,
+                          const Scalar<double>& det_spatial_metric,
+                          const tnsr::i<double, Dim>& unit_normal) noexcept {
+  Matrix result(Dim + 2, Dim + 2);
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest moving this down to line ~ 136 above its first use

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 12]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+  const double h_w_minus_one =
+      get(lorentz_factor) *
+      (get(specific_internal_energy) + get(pressure) / get(rest_mass_density) +
+       get(lorentz_factor) * velocity_squared / (get(lorentz_factor) + 1.0));
+
+  double sound_speed_over_d =
+      sqrt(1.0 - velocity_squared * get(sound_speed_squared) -
+           square(normal_velocity) * (1.0 - get(sound_speed_squared)));
+  sound_speed_over_d *= get(lorentz_factor);
+  sound_speed_over_d = sqrt(get(sound_speed_squared)) / sound_speed_over_d;
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest the simpler to read
```
const double d = ...;
const double sound_speed_over_d = sqrt(get(sound_speed_squared)) / d;
```

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 13]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+        result(i + 2, vector_id) =
+            get(specific_enthalpy) *
+            (tangent_oneform.get(i) +
+             local_prefactor * spatial_velocity_oneform.get(i));
+      }
+    };
+    degenerate_eigenvector(1, unit_tangent_oneform);
+
+    if (Dim > 2) {
+      make_overloader([](std::integral_constant<size_t, 1> /*dim*/,
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp:202]: This complexity is surprising at first.. I suggest adding a short comment (above this line) similar to

"Here we use an overload of `orthonormal_oneform` that is only defined in 3D, so it must be called from within a `make_overloader`."

🙁

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 14]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+      get(kappa_over_density) * h_w_minus_one -
+      specific_enthalpy_times_lorentz_factor * get(sound_speed_squared);
+
+  const double prefactor = specific_enthalpy_times_lorentz_factor *
+                           (get(kappa_over_density) - get(sound_speed_squared));
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = prefactor * spatial_velocity_oneform.get(i);
+  }
+
+  // if Dim > 1 make degenerate eigenvectors
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: make -> set, for consistency with comments above

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 15]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+  if (Dim > 1) {
+    const double two_h_w_minus_one =
+        h_w_minus_one + specific_enthalpy_times_lorentz_factor;
+    const auto unit_tangent_oneform =
+        orthonormal_oneform(unit_normal, inv_spatial_metric);
+    const auto degenerate_eigenvector =
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: [nitpick] I suggest a more "active" name to clarify what this is doing... perhaps "set_degenerate_eigenvector"? ... up to you on this one.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 16]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+                         const Scalar<double>& pressure,
+                         const Scalar<double>& specific_enthalpy,
+                         const Scalar<double>& kappa_over_density,
+                         const Scalar<double>& sound_speed_squared,
+                         const Scalar<double>& lorentz_factor,
+                         const tnsr::ii<double, Dim>& spatial_metric,
+                         const tnsr::II<double, Dim>& inv_spatial_metric,
+                         const Scalar<double>& det_spatial_metric,
+                         const tnsr::i<double, Dim>& unit_normal) noexcept {
+  Matrix result(Dim + 2, Dim + 2);
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto move down, to first use

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 17]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+      inv_specific_enthalpy / get(sound_speed_squared);
+  result(eigenvector_id, 0) = inv_h_c_squared * h_minus_w;
+  result(eigenvector_id, 1) = inv_h_c_squared * get(lorentz_factor);
+  for (size_t i = 0; i < Dim; ++i) {
+    result(eigenvector_id, i + 2) =
+        result(eigenvector_id, 1) * spatial_velocity.get(i);
+  }
+  result(eigenvector_id, 1) *= -1.0;
+
+  // if Dim > 1 make degenerate eigenvectors
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto "make -> set"

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 18]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+  if (Dim > 1) {
+    const double prefactor = -inv_specific_enthalpy / one_minus_vn_squared;
+    const auto unit_tangent_oneform =
+        orthonormal_oneform(unit_normal, inv_spatial_metric);
+    const auto degenerate_eigenvector =
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: ditto name of lambda

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 19]
```diff
@@ -87,20 +92,322 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+      const double local_prefactor = result(vector_id, 0) * normal_velocity;
+      for (size_t i = 0; i < Dim; ++i) {
+        result(vector_id, i + 2) = inv_specific_enthalpy * unit_tangent.get(i) -
+                                   local_prefactor * unit_normal_vector.get(i);
+      }
+    };
+    degenerate_eigenvector(1, unit_tangent_oneform);
+
+    if (Dim > 2) {
+      make_overloader([](std::integral_constant<size_t, 1> /*dim*/,
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp:352]: ditto comment on complexity

**Review by fmahebert — COMMENTED** [review]
Thanks @carmaza, I know my review was asking for some large changes in the test.

I'm happy with all your updates, so please go ahead and squash (also rebase to get the helpers that I had cherry-picked in a different PR) - up to you whether to do this now or after the other dependency is merged.

Below are a few minor nitpicks and comments regarding clarity; please squash these in directly while you squash/rebase.

**Comment by teukolsky** [comment]
I've checked the equations in the documentation - look fine.

**Comment by carmaza** [comment]
Squashed and rebased on develop.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp [hunk 20]
```diff
@@ -87,20 +92,324 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+        result(i + 2, vector_id) =
+            get(specific_enthalpy) *
+            (tangent_oneform.get(i) +
+             local_prefactor * spatial_velocity_oneform.get(i));
+      }
+    };
+    set_degenerate_eigenvector(1, unit_tangent_oneform);
+
+    if (Dim > 2) {
+      // orthonormal_form for two forms is defined for 3-d only
```
Comment (fmahebert) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp]: `orthonormal_form` -> `orthonormal_oneform`

(here and in similar comment below)

**Review by fmahebert — COMMENTED** [review]
One small typo, otherwise LGTM! 👍 

Squash directly.

**Review by fmahebert — DISMISSED** [review]

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 21]
```diff
@@ -74,8 +75,143 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+ *
+ * The principal symbol of the Valencia formulation, \f$A\f$, is diagonalizable,
+ * allowing the introduction of characteristic variables \f$U_\text{char}\f$.
+ * In terms of the conservative variables \f$U\f$, \f$U = RU_\text{char}\f$ and
+ * \f$U_\text{char} = LU\f$, where \f$R\f$ is the matrix whose columns are the
+ * right eigenvectors of \f$A\f$, and \f$L = R^{-1}\f$ contains the left
+ * eigenvectors as rows. Here we take the ordering
+ * \f$U = [\tilde D, \tilde \tau, \tilde S_i]^T\f$. Explicitly,
+ * \f$L = [L_-\, L_1\, L_2\, L_3\, L_+]^T\f$, where [Teukolsky, to be published]
```
Comment (kidder) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: why are we not citing a published paper?

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp [hunk 22]
```diff
@@ -74,8 +75,143 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
...
+ *
+ * \f{align*}
+ * L_\pm &= \dfrac{1}{2h W c_s^2(1 - v_n^2)}\left[\begin{array}{ccc}
```
Comment (kidder) [inline src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:162]: can you either add commas between the elements or write this as a column vector with a `^T` to indicate transpose.  We find the current format hard to parse

**Review by kidder — COMMENTED** [review]
reviewed with @nilsdeppe

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp (hunk 22 above)
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:162]: Added the commas to all left eigenvectors.

File: src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp (hunk 21 above)
Comment (carmaza) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: My understanding is that no other author has worked out this form of the eigenvectors in terms of an arbitrary normal orientation.

Comment (teukolsky) [reply src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp:159]: That's correct - these expressions are much simpler than those in the literature. They probably won't get published till I've finished the MHD ones, maybe a few months.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@fmahebert can you please let us know if you're happy with this when you get a chance? Thanks!

Edit: I was too slow, sorry!
